Add findmin, findmax, argmin, and argmax

sethaxen opened the pull request:
This PR adds versions of `findmin`, `findmax`, `argmin` and `argmax` that ignore `NaN`s in the codomain. It does not include methods that take a `dims` keyword.

Fixes #31

A simple benchmark:
```julia
julia> using NaNMath, BenchmarkTools

julia> x = map(x -> x < 0.9 ? x : NaN, rand(1_000, 1_000));

julia> @btime findmax($x);
  1.914 ms (0 allocations: 0 bytes)

julia> @btime $(NaNMath.findmax)($x);
  2.809 ms (0 allocations: 0 bytes)

julia> @btime $(NaNMath.maximum)($x);
  2.100 ms (0 allocations: 0 bytes)

julia> @btime findmax(cos, $x);
  9.590 ms (0 allocations: 0 bytes)

julia> @btime $(NaNMath.findmax)(cos, $x);
  9.226 ms (0 allocations: 0 bytes)

julia> @btime argmax($x);
  1.915 ms (0 allocations: 0 bytes)

julia> @btime $(NaNMath.argmax)($x);
  2.725 ms (0 allocations: 0 bytes)

julia> @btime argmax(cos, $x);
  8.348 ms (0 allocations: 0 bytes)

julia> @btime $(NaNMath.argmax)(cos, $x);
  8.560 ms (0 allocations: 0 bytes)
```

mlubin commented on src/NaNMath.jl:
```diff
@@ -354,4 +354,164 @@ for f in (:min, :max)
+* `domain`: A non-empty iterable of floating point numbers or `Missing`.
+
+##### Returns:
+* Returns a pair of a value in the codomain (outputs of `f`, defaulting to `identity`) and
```
The default value of `f` should be mentioned in the description of the argument, not in the "Returns" section

mlubin commented on src/NaNMath.jl:
```diff
@@ -354,4 +354,164 @@ for f in (:min, :max)
+findmax(x) = findmax(identity, x)
+
+function _findminmax_op(cmp)
+    return (x1_i1, x2_i2) -> begin
```
Please use more explicit variable names. A reader might have trouble guessing what `x1_i1` means.

mlubin commented on src/NaNMath.jl:
```diff
@@ -354,4 +354,164 @@ for f in (:min, :max)
+function _findminmax_op(cmp)
+    return (x1_i1, x2_i2) -> begin
+        x1 = first(x1_i1)
+        x1 === missing && return x1_i1
```
The docstring above doesn't mention how `missing` is handled.

mlubin commented on src/NaNMath.jl:
```diff
@@ -354,4 +354,164 @@ for f in (:min, :max)
+* `domain`: A non-empty iterable of floating point numbers or `Missing`.
+
+##### Returns:
+* Returns a value `x` in the domain of `f` for which `f(x)` is minimised. If there are
```
Nit: minimized (findmax above uses the american spelling).

mlubin reviewed (commented):
Thanks for the contribution. I could accept this after some reworking.

mlubin commented:
Please rebase on top of https://github.com/mlubin/NaNMath.jl/pull/54. I saw the 1.0 build is failing, so I dropped support for 1.0.

sethaxen commented:
@mlubin thanks for the review! I have addressed all of your suggestions. Because this PR adds new functions to the API, I incremented the minor version, as specified by the [Pkg.jl docs](https://pkgdocs.julialang.org/v1/toml-files/#The-version-field).

mlubin commented:
It looks like tests are failing on 1.6.

sethaxen commented:
> It looks like tests are failing on 1.6.

Should be fixed now. A few tests were testing against Base methods not introduced until Julia 1.7 or hitting Base bugs on 1.6 fixed on 1.7.

sethaxen commented:
@mlubin would you like to you re-review?

mlubin commented on test/runtests.jl:
```diff
@@ -70,3 +70,117 @@ using Test
 @test NaNMath.max(NaN, NaN, 0.0, 1.0) == 1.0
+
+@testset "findmin/findmax" begin
+    xvals = [
```
The scope of `xvals` could be tighter. I'd recommend:
```julia
if VERSION >= v"1.7"
   xvals = [...]
   @test ...
end
```
Same below.

mlubin reviewed (commented):
Thanks for the reminder. The new version looks good. I have a minor comment on the tests. Please also update the README to mention the new functions.

sethaxen commented:
Done!

mlubin commented:
Hmm, there's a test failure on 1.6.

jd-foster commented:
The failure on x86 but not x64 is likely due to a different coincidental ordering of `Dict` keys, since this occurs on x86 on the PR branch:
```julia
julia> NaNMath.argmin(Dict(:x => 3, :w => NaN, :z => -1.0, :y => NaN))
:w

julia> NaNMath.argmax(Dict(:x => 3, :w => NaN, :z => -1.0, :y => NaN))
:w
```

ChrisRackauckas commented:
Closing for https://github.com/JuliaMath/NaNMath.jl/pull/62/
